tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+#                     _r.memoryformatOptional(2))
+#
+#    Where does 'names' come from? It involves special local init:
+#
+#      auto __names = _r.toDimnameListOptional(1);
+#      c10::optional<DimnameList> names =
+#          __names ? c10::make_optional(DimnameList(__names.value()))
+#                  : c10::nullopt;
+#
+#    Where does 'options' come from? It involves special local init
+#    for TensorOptions. Note that Python side has the additional
+#    'requires_grad' field:
+#
+#      const auto options = TensorOptions()
+#          .dtype(_r.scalartype(3))
+#          .device(_r.device(5))
+#          .layout(_r.layoutOptional(4))
+#          .requires_grad(_r.toBool(7))
+#          .pinned_memory(_r.toBool(6));
+#
+#    In some other cases one Python Argument can map to multiple C++
+#    Arguments. For example:
+#
+#     aten::max.names_dim(Tensor self, Dimname dim, bool keepdim=False)
+#       -> (Tensor values, Tensor indices)
+#
+#            Python Args               Cpp Args          Binding Exprs
+#     ---------------------------------------------------------------------
+#                               +----> max               'out[0]'
+#                              /-----> max_values        'out[1]
+#         0: input            /        self              '_r.tensor(0)'
+#         1: dim             /         dim               '_r.dimname(1)'
+#         2: keepdim        /          keepdim           '_r.toBool(2)'
+#         3: out      -----+           [local init] out  '_r.tensorlist_n<2>(3)'
+#
+#    As demonstrated above, the binding can involve reordering,
+#    packing, unpacking and special local inits.
+#
+# The following diagram illustrates how different data models are used
+# to generate different parts of the code:
+#
+#             +-------------------+     +-------------------------------------+
+#   +-------> | Python Signatures | --> |       PythonArgParser Schema        |
+#   |         +-------------------+     +-------------------------------------+
+#   |                         |                            .
+#   |                         |                            .
+#   |                         |                            .
+# +------------------+        |         +-------------------------------------+
+# | Native Functions |        +-------> | PythonArgParser -> Cpp Args Binding |
+# +------------------+        |         +-------------------------------------+
+#   |                         |                            .
+#   |                         |                            .
+#   |                         |                            .
+#   |         +-------------------+     +-------------------------------------+
+#   +-------> |  Cpp Signatures   | --> |        Cpp Function Dispatch        |
+#             +-------------------+     +-------------------------------------+
+
+
+@dataclass(frozen=True)
+class PythonArgument:
+    name: str
+    type: Type
```

Comment:
This sort of suggests that maybe we should have kept a reference to the full `Argument` here. Not sure though! Going to keep reading.